File: python/eval/adhoc.py
#!/usr/bin/env python3
# Copyright 2026 Jeremy Lilley (jeremy@jlilley.net)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Ad-hoc language detection of a single phrase from the command line.

Usage:
    python3 adhoc.py --languages efigs --phrase "this is a test sentence"
    python3 adhoc.py --languages europe_west_common --phrase "Bonjour le monde"
"""

import argparse
import sys

from langidentify import Detector, Language, Model


def main():
  parser = argparse.ArgumentParser(
    description="Detect the language of a single phrase.")
  parser.add_argument(
    "--languages", required=True,
    help="Comma-separated language codes or group aliases (e.g. efigs, europe_west_common)")
  parser.add_argument(
    "--phrase", required=True,
    help="The phrase to detect")
  parser.add_argument(
    "--model", choices=["lite", "full", "best"], default="best",
    help="Model variant: lite, full, or best (auto-detect, default)")
  args = parser.parse_args()

  languages = Language.from_comma_separated(args.languages)
  if not languages:
    print("No languages matched.", file=sys.stderr)
    sys.exit(1)

  if args.model == "lite":
    model = Model.load_lite(languages)
  elif args.model == "full":
    model = Model.load_full(languages)
  else:
    model = Model.load(languages)
  detector = Detector(model)

  result = detector.detect(args.phrase)
  results = detector.results

  print(f"Result: {result.iso_code} ({result.name.replace('_', ' ').title()})")
  print(f"Gap:    {results.gap:.4f}")
  print()

  # Show per-language score breakdown, sorted by total score (best first).
  langs = model.languages
  scores = results.scores
  total = results.total_scores

  # Build (total_score, lang_index) pairs sorted best first.
  order = sorted(range(len(langs)), key=lambda i: total[i], reverse=True)

  print(f"{'lang':>6}  {'ngram':>10}  {'hits':>5}  {'tw':>10}  {'tw_hits':>7}  {'total':>10}")
  print(f"{'----':>6}  {'-----':>10}  {'----':>5}  {'--':>10}  {'------':>7}  {'-----':>10}")
  for li in order:
    print(
      f"{langs[li].iso_code:>6}  "
      f"{scores.ngram_scores[li]:>10.3f}  "
      f"{scores.ngram_hits_per_lang[li]:>5d}  "
      f"{scores.tw_scores[li]:>10.3f}  "
      f"{scores.tw_hits_per_lang[li]:>7d}  "
      f"{total[li]:>10.3f}")


if __name__ == "__main__":
  main()

File: rust/eval/src/adhoc.rs
// Copyright 2026 Jeremy Lilley
// Licensed under the Apache License, Version 2.0

//! Ad-hoc language detection of a single phrase from the command line.
//!
//! Usage:
//!     cargo run --bin adhoc -- --languages efigs --phrase "this is a test sentence"
//!     cargo run --bin adhoc -- --languages europe_west_common --phrase "Bonjour le monde"

use langidentify::{Detector, Language};
use langidentify_eval::common::{load_model, ModelVariant};
use std::process;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let (languages_str, phrase, variant) = parse_args(&args);

    let languages = match Language::from_comma_separated(&languages_str) {
        Ok(langs) => langs,
        Err(e) => {
            eprintln!("Invalid languages: {}", e);
            process::exit(1);
        }
    };

    let model = load_model(variant, &languages).unwrap_or_else(|e| {
        eprintln!("Failed to load model: {}", e);
        process::exit(1);
    });

    let mut detector = Detector::new(model.clone());
    let result = detector.detect(&phrase);
    let results = detector.results();

    println!("Result: {} ({:?})", result.iso_code(), result);
    println!("Gap:    {:.4}", results.gap);
    println!();

    let langs = model.languages();
    let scores = &results.scores;
    let total = &results.total_scores;

    // Sort by total score descending.
    let mut order: Vec<usize> = (0..langs.len()).collect();
    order.sort_by(|&a, &b| total[b].partial_cmp(&total[a]).unwrap());

    println!(
        "{:>6}  {:>10}  {:>5}  {:>10}  {:>7}  {:>10}",
        "lang", "ngram", "hits", "tw", "tw_hits", "total"
    );
    println!(
        "{:>6}  {:>10}  {:>5}  {:>10}  {:>7}  {:>10}",
        "----", "-----", "----", "--", "------", "-----"
    );
    for li in order {
        println!(
            "{:>6}  {:>10.3}  {:>5}  {:>10.3}  {:>7}  {:>10.3}",
            langs[li].iso_code(),
            scores.ngram_scores[li],
            scores.ngram_hits_per_lang[li],
            scores.tw_scores[li],
            scores.tw_hits_per_lang[li],
            total[li],
        );
    }
}

fn parse_args(args: &[String]) -> (String, String, ModelVariant) {
    let mut languages = None;
    let mut phrase = None;
    let mut variant = ModelVariant::Best;
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--languages" => {
                i += 1;
                languages = Some(args[i].clone());
            }
            "--phrase" => {
                i += 1;
                phrase = Some(args[i].clone());
            }
            "--model" => {
                i += 1;
                variant = ModelVariant::from_str(&args[i]);
            }
            "--help" | "-h" => {
                eprintln!("Usage: adhoc --languages <langs> --phrase <text> [--model lite|full|best]");
                process::exit(0);
            }
            _ => {
                eprintln!("Unknown argument: {}", args[i]);
                process::exit(1);
            }
        }
        i += 1;
    }
    let languages = languages.unwrap_or_else(|| {
        eprintln!("--languages is required");
        process::exit(1);
    });
    let phrase = phrase.unwrap_or_else(|| {
        eprintln!("--phrase is required");
        process::exit(1);
    });
    (languages, phrase, variant)
}
